tests/src/python/qgis_wrapped_server.py

```diff
@@ -100,6 +108,42 @@ def responseComplete(self):
     qgs_server.serverInterface().registerFilter(filter)
 
 
+def num2deg(xtile, ytile, zoom):
+    """This returns the NW-corner of the square. Use the function with xtile+1 and/or ytile+1 
+    to get the other corners. With xtile+0.5 & ytile+0.5 it will return the center of the tile."""
+    n = 2.0 ** zoom
+    lon_deg = xtile / n * 360.0 - 180.0
+    lat_rad = math.atan(math.sinh(math.pi * (1 - 2 * ytile / n)))
+    lat_deg = math.degrees(lat_rad)
+    return (lat_deg, lon_deg)
+
+
+class XYZFilter(QgsServerFilter):
+    """XYZ server, example: ?MAP=/path/to/projects.qgs&SERVICE=XYZ&X=1&Y=0&Z=1&LAYERS=world"""
+
+    def requestReady(self):
+        handler = self.serverInterface().requestHandler()
+        if handler.parameter('SERVICE') == 'XYZ':
+            x = int(handler.parameter('X'))
+            y = int(handler.parameter('Y'))
+            z = int(handler.parameter('Z'))
+            # NW corner
+            lat_deg, lon_deg = num2deg(x, y, z)
+            # SE corner
+            lat_deg2, lon_deg2 = num2deg(x + 1, y + 1, z)
+            handler.setParameter('SERVICE', 'WMS')
+            handler.setParameter('REQUEST', 'GetMap')
+            handler.setParameter('VERSION', '1.3.0')
+            handler.setParameter('SRS', 'EPSG:4326')
+            handler.setParameter('HEIGHT', '256')
+            handler.setParameter('WIDTH', '256')
+            handler.setParameter('BBOX', "{},{},{},{}".format(lat_deg2, lon_deg, lat_deg, lon_deg2))
+
+
+xyzfilter = XYZFilter(qgs_server.serverInterface())
+qgs_server.serverInterface().registerFilter(xyzfilter)
```

Comment:
@dmarteau could the XYZ service be a QGIS Server Service instead of QGIS Server FIlter ?

Reply:
As long it obey OGC service rules for services (i.e use SERVICE query parameter) it can be implemented as a service of its own: but functionally, this is just a frontend to a WMS service so, imho, this is more suited as a filter plugin.

Reply:
That would be a nice idea for the future indeed!
But for now, this is just a quick and dirty (but working) filter-plugin needed for testing QGIS Desktop xyz (really wms) provider with different authentication plugins in multi-threading mode: this is not a test for QGIS Server.

This wrapped_server is really just a useful live endpoint for integration tests on Travis and locally ...

I though it was useful to share, because this wrapper script is very useful when you need a to quickly stand up a testing enpoint that you can fully control (there is even an OAuth2 enabled version buried in an old PR).